File: practitioner-code/payments-java/src/main/java/com/example/payments/application/WalletService.java
package com.example.payments.application;

import com.example.payments.domain.Currency;
import com.example.payments.domain.Money;
import com.example.payments.domain.Wallet;
import com.example.payments.domain.WalletId;
import com.example.payments.domain.WalletRepository;
import com.example.payments.domain.exception.OptimisticLockException;

import java.math.BigDecimal;
import java.util.NoSuchElementException;

/**
 * APPLICATION SERVICE (hexagonal architecture). Orchestrates use cases.
 *
 * What this layer DOES:
 *   - Load aggregates by id (via the repository port).
 *   - Call domain methods to enforce invariants (debit/credit).
 *   - Persist via the repository.
 *   - Handle OCC conflicts with bounded retry.
 *   - Translate "wallet not found" into the application's language.
 *
 * What this layer DOES NOT do:
 *   - Business rules (those live in Wallet/Money).
 *   - Know about HTTP, JSON, JPA, or any framework. Inputs are domain
 *     types. Part 5 wires a Spring controller on top - controllers translate
 *     HTTP <-> these method calls, nothing more.
 *
 * Why retry here (not in domain, not in repository): retry is an
 * APPLICATION POLICY (how hard to try). The domain is pure rules; the
 * repository reports facts ("conflict"); the application decides what to do.
 */
public class WalletService {

    private final WalletRepository wallets;
    private final int maxRetries;

    public WalletService(WalletRepository wallets) {
        this(wallets, 3);   // sane production default
    }

    public WalletService(WalletRepository wallets, int maxRetries) {
        // Dependency Injection by constructor: explicit, easy to test, no
        // hidden globals. The application service depends on the PORT, not
        // on any concrete repository (DIP, the D in SOLID). maxRetries is
        // a POLICY, also injected - tests can use a higher value to ride
        // out heavy contention deterministically.
        this.wallets = wallets;
        if (maxRetries < 1) throw new IllegalArgumentException("maxRetries must be >= 1");
        this.maxRetries = maxRetries;
    }

    public WalletId openWallet(String openingAmount, Currency currency) {
        WalletId id = WalletId.newId();
        Wallet w = Wallet.open(id, new Money(new BigDecimal(openingAmount), currency));
        wallets.save(w);
        return id;
    }

    /** Debit the wallet. Retries up to MAX_RETRIES times on OCC conflicts. */
    public Money debit(WalletId id, Money amount) {
        return withOptimisticRetry(() -> {
            Wallet w = loadOrThrow(id);
            w.debit(amount);
            wallets.save(w);
            return w.balance();
        });
    }

    public Money credit(WalletId id, Money amount) {
        return withOptimisticRetry(() -> {
            Wallet w = loadOrThrow(id);
            w.credit(amount);
            wallets.save(w);
            return w.balance();
        });
    }

    public Money balance(WalletId id) {
        return loadOrThrow(id).balance();
    }

    // ---- internals --------------------------------------------------------
    private Wallet loadOrThrow(WalletId id) {
        return wallets.findById(id).orElseThrow(
                () -> new NoSuchElementException("wallet not found: " + id.value()));
    }

    private <T> T withOptimisticRetry(java.util.function.Supplier<T> action) {
        OptimisticLockException last = null;
        for (int attempt = 0; attempt < maxRetries; attempt++) {
            try { return action.get(); }
            catch (OptimisticLockException e) { last = e; }
        }
        throw last;   // give up; let the caller (e.g. an HTTP layer) translate to 409
    }
}

File: practitioner-code/payments-python/src/payments/application/wallet_service.py
"""
APPLICATION SERVICE (hexagonal architecture). Orchestrates use cases.

What this layer DOES:
  - Load aggregates by id (via the repository port).
  - Call domain methods to enforce invariants (debit/credit).
  - Persist via the repository.
  - Handle OCC conflicts with bounded retry.
  - Translate "wallet not found" into the application's language.

What this layer DOES NOT do:
  - Business rules (those live in Wallet/Money).
  - Know about HTTP, JSON, SQLAlchemy, or any framework. Inputs are domain
    types. Part 5 wires a FastAPI router on top - the router translates
    HTTP <-> these method calls, nothing more.

Why retry here (not in domain, not in repository): retry is an APPLICATION
POLICY (how hard to try). The domain is pure rules; the repository reports
facts ("conflict"); the application decides what to do.
"""
from __future__ import annotations

from decimal import Decimal
from typing import Callable, TypeVar

from payments.domain.currency import Currency
from payments.domain.exceptions import OptimisticLockError
from payments.domain.money import Money
from payments.domain.repository import WalletRepository
from payments.domain.wallet import Wallet
from payments.domain.wallet_id import WalletId

T = TypeVar("T")


class WalletNotFoundError(Exception):
    """Application-level: this id has no wallet (translate to HTTP 404 later)."""


class WalletService:
    def __init__(self, wallets: WalletRepository, max_retries: int = 3) -> None:
        # Dependency Injection via constructor: explicit, easy to test, no
        # hidden globals. Depends on the PORT (abstract), not on any
        # concrete repo (DIP).  max_retries is a POLICY, also injected.
        if max_retries < 1:
            raise ValueError("max_retries must be >= 1")
        self._wallets = wallets
        self._max_retries = max_retries

    def open_wallet(self, opening_amount: str, currency: Currency) -> WalletId:
        wid = WalletId.new()
        w = Wallet.open(wid, Money(Decimal(opening_amount), currency))
        self._wallets.save(w)
        return wid

    def debit(self, wallet_id: WalletId, amount: Money) -> Money:
        def action() -> Money:
            w = self._load_or_raise(wallet_id)
            w.debit(amount)
            self._wallets.save(w)
            return w.balance
        return self._with_optimistic_retry(action)

    def credit(self, wallet_id: WalletId, amount: Money) -> Money:
        def action() -> Money:
            w = self._load_or_raise(wallet_id)
            w.credit(amount)
            self._wallets.save(w)
            return w.balance
        return self._with_optimistic_retry(action)

    def balance(self, wallet_id: WalletId) -> Money:
        return self._load_or_raise(wallet_id).balance

    # ---- internals --------------------------------------------------------
    def _load_or_raise(self, wallet_id: WalletId) -> Wallet:
        w = self._wallets.find_by_id(wallet_id)
        if w is None:
            raise WalletNotFoundError(f"wallet not found: {wallet_id.value}")
        return w

    def _with_optimistic_retry(self, action: Callable[[], T]) -> T:
        last: OptimisticLockError | None = None
        for _ in range(self._max_retries):
            try:
                return action()
            except OptimisticLockError as e:
                last = e
        assert last is not None
        raise last   # give up; let the caller (e.g. HTTP layer) translate to 409
